lib/Dialect/Comb/Transforms/IntRangeOptimizations.cpp

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/Comb/CombPasses.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+#include "mlir/Analysis/DataFlowFramework.h"
+#include "mlir/Dialect/Arith/Transforms/Passes.h"
+
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Analysis/DataFlow/IntegerRangeAnalysis.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace circt;
+using namespace circt::comb;
+using namespace mlir;
+using namespace mlir::dataflow;
+
+namespace circt {
+namespace comb {
+#define GEN_PASS_DEF_COMBINTRANGENARROWING
+#include "circt/Dialect/Comb/Passes.h.inc"
+} // namespace comb
+} // namespace circt
+
+/// Gather ranges for all the values in `values`. Appends to the existing
+/// vector.
+static LogicalResult collectRanges(DataFlowSolver &solver, ValueRange values,
+                                   SmallVectorImpl<ConstantIntRanges> &ranges) {
+  for (Value val : values) {
+    auto *maybeInferredRange =
+        solver.lookupState<IntegerValueRangeLattice>(val);
+    if (!maybeInferredRange || maybeInferredRange->getValue().isUninitialized())
+      return failure();
+
+    const ConstantIntRanges &inferredRange =
+        maybeInferredRange->getValue().getValue();
+    ranges.push_back(inferredRange);
+  }
+  return success();
+}
+
+namespace {
+template <typename CombOpTy>
+struct CombOpNarrow : public OpRewritePattern<CombOpTy> {
+  CombOpNarrow(MLIRContext *context, DataFlowSolver &s)
+      : OpRewritePattern<CombOpTy>(context), solver(s) {}
+
+  LogicalResult matchAndRewrite(CombOpTy op,
+                                PatternRewriter &rewriter) const override {
+
+    auto opWidth = op.getType().getIntOrFloatBitWidth();
+    if (op->getNumOperands() != 2 || op->getNumResults() != 1)
+      return rewriter.notifyMatchFailure(
+          op, "Only support binary operations with one result");
+
+    SmallVector<ConstantIntRanges> ranges;
+    if (failed(collectRanges(solver, op->getOperands(), ranges)))
+      return rewriter.notifyMatchFailure(op, "input without specified range");
+    if (failed(collectRanges(solver, op->getResults(), ranges)))
+      return rewriter.notifyMatchFailure(op, "output without specified range");
+
+    auto removeWidth = ranges[0].umax().countLeadingZeros();
+    for (const ConstantIntRanges &range : ranges) {
+      auto rangeCanRemove = range.umax().countLeadingZeros();
+      removeWidth = std::min(removeWidth, rangeCanRemove);
+    }
+    if (removeWidth == 0)
+      return rewriter.notifyMatchFailure(op, "no bits to remove");
+    else if (removeWidth == opWidth)
+      return rewriter.notifyMatchFailure(
+          op, "all bits to remove - replace by zero");
```

Comment:
Nit: unnecessary else after return
```suggestion
    if (removeWidth == 0)
      return rewriter.notifyMatchFailure(op, "no bits to remove");
    
    if (removeWidth == opWidth)
      return rewriter.notifyMatchFailure(
          op, "all bits to remove - replace by zero");
```